### PR TITLE
Introduce single-node discovery

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds a single node discovery type. With this discovery type, a node will elect itself as master and never form a cluster with another node.

### CODE REVIEW

**Comment by jasontedor** [comment]
A follow-up will reveal the true purpose of this. :wink:

File: core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Override
+    public void publish(final ClusterChangedEvent event, final AckListener listener) {
+
+    }
+
+    @Override
+    public DiscoveryStats stats() {
+        return null;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: I'm doubting if this should returns a `new DiscoveryStats(null)`. wdyt?

File: core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                    }
+
+                    @Override
+                    public boolean runOnlyOnMaster() {
+                        return false;
+                    }
+
+                };
+        final ClusterStateTaskConfig config = ClusterStateTaskConfig.build(Priority.URGENT);
+        clusterService.submitStateUpdateTasks(
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: out of curiosity - why didn't you use a simple `ClusterStateUpdateTask` ?

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java [hunk 3]
```diff
@@ -474,7 +474,6 @@ protected void sendPings(final TimeValue timeout, final PingingRound pingingRoun
             Stream.concat(pingingRound.getSeedNodes().stream(), nodesFromResponses.stream())
                 .collect(Collectors.toMap(DiscoveryNode::getAddress, Function.identity(), (n1, n2) -> n1));
 
-
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: nit: unneeded?

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings
+                .builder()
+                .put(super.nodeSettings(nodeOrdinal))
+                .put("discovery.type", "single-node")
+                .build();
+    }
+
+    public void testDoesNotPing() throws IOException, InterruptedException, ExecutionException {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java]: I'm a bit confused by this test - as far as I can tell it doesn't do anything with single node discovery? what are we testing here?

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryTests.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            final MockTransportService transportService =
+                    MockTransportService.createNewService(settings, version, threadPool, null);
+            stack.push(transportService);
+            transportService.start();
+            final DiscoveryNode node = transportService.getLocalNode();
+            final ClusterService clusterService = createClusterService(threadPool, node);
+            stack.push(clusterService);
+            final SingleNodeDiscovery discovery =
+                    new SingleNodeDiscovery(Settings.EMPTY, clusterService);
+            discovery.startInitialJoin();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryTests.java]: doesn't this go async? should we wait via an observer? maybe take the code from node.java ?

**Review by bleskes — CHANGES_REQUESTED** [review]
Main change looks good to me. Left some minor comments.

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java (hunk 4 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java]: A single node with discovery type set to `single-node` is started, and then we attempt to ping as Zen discovery would. The purpose is to test that the node will not join a cluster with Zen discovery configured. Does that make sense? Do you have an idea for a better test? Do you think that we need to test that two nodes with `single-node` configured as the discovery type will not join each other?

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryTests.java (hunk 5 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryTests.java]: Good catch. I pushed 5caab7d12d367f33eaeca9ab8d41ee76cfcdce14.

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: I pushed f235c8bdc7dfb8dc9ca5edc9878d5808cfb7feed.

File: core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: No reasoning other than this is what my fingers did.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: No, it wasn't my fingers. I see why I did this. A `ClusterStateUpdateTask` can not override `runOnlyOnMaster` which we need to set to `false` here. 😄

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java (hunk 4 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java]: Oh, I see. So it's more like a "node has a good cloaking device" test. Reading the name I assumed you check that a single-node discovery doesn't ping other nodes (i.e., doesn't join). I think this test is nice (with a new name). Maybe another good test is to start two nodes with the same cluster name and port ranges that would normally cause them to find each other and see that they don't (it's a bit tricky w.r.t fixing ports but I think we can figure something out)

File: core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: I pushed e10a245abc91dbb89d8c095d7d6bed1fbefa950e.

**Comment by jasontedor** [comment]
Thanks @bleskes; I have responded to your feedback.

File: core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/single/SingleNodeDiscovery.java]: Oh right. I forgot about that. ++

**Review by bleskes — APPROVED** [review]
LGTM. See if you want to bite your teeth on that test suggestion

File: core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java (hunk 4 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/single/SingleNodeDiscoveryIT.java]: @bleskes I pushed 4c913108aac79a8cf42b05243c803eddbe159d2e to start a discussion. Let me know what you think.

**Comment by jasontedor** [comment]
Thanks @bleskes.

**Comment by chitraj8** [comment]
I am trying to use elasticsearch version 5.6 and will need test it in a single node type. please suggest the configuration files that will require the change and get started.

**Comment by jasontedor** [comment]
@chitraj8 Please check the docs, and please use the [forum](https://discuss.elastic.co) if you need more help. We reserve GitHub issues for verified bug reports and feature requests, and merged pull requests are not the place to ask general questions.

**Pull request merged**